helix-core/src/main/java/org/apache/helix/cloud/azure/AzureHttpUtil.java

```diff
@@ -0,0 +1,45 @@
+package org.apache.helix.cloud.azure;
+
+import javax.net.ssl.SSLException;
+import org.apache.helix.HelixCloudProperty;
+import org.apache.http.client.HttpRequestRetryHandler;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.protocol.HttpContext;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.net.UnknownHostException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A dedicated http client for retrieving information from Azure Instance Metadata Service
+ */
+class AzureHttpUtil {
```

Comment:
On general question, Yi has added one httpclient in the Helix-rest for instance health check. Can we just leverage that one?
In that case, just move the code to helix-core, or better to helix-common (I hope we have one), then leverage that client.

Reply:
I've checked the current custom rest client. It's not easy to adapt it to our use case, specifically for those user defined parameter input, such as max retry, timeout. Also the separate http client I have is easier to mock in the unit test.

Reply:
httpclient should be a fundamental component. If we have 2 of them in our codebase, it would be confusing and harder for us to maintain.
How difficult will it be to merge 2 use cases? Please feel free to refactor if necessary.

@i3wangyi Can you please comment as well?

Reply:
Yes. The current CustomRestClient has one instance of `HttpClient`, a parent class of the `CloseableHttpClient` the method tries to return. However, I see the AzureHttpUtil is in helix-core while that factory is only visible in helix-rest only (because previously there's no need for helix-core to perform HTTP request).

Essentially, the factory pattern is similar and we can extract the core part of it to helix-core and leave the difference in the separate module. For example, in helix-core, create one HttpClientFactory which accepts a set of general parameters (timeout, ssl, retry..) without respecting HelixCloudProperty; And use it in rest as well as rest of places in helix-core

Reply:
It is OK if we move the current client class to the helix-core now. Since there is real usage of it.

Reply:
Actually I double checked and found that I can just remove the util function. The processor itself can just directly make the http call, and passing the information it needs to pass. So the future usage should be, any class that needs to make the call would directly call getHttpClient, providing whatever config they would like to pass in.